flake8_trio.py

```diff
@@ -255,6 +255,135 @@ def check_for_trio100(self, node: Union[ast.With, ast.AsyncWith]) -> None:
+                )
+
+        self.unraised = outer_unraised
+
+    def visit_Raise(self, node: ast.Raise):
+        # if there's an exception that must be raised
+        # and none of the valid ways of re-raising it is done
+        if self.unraised and not (
+            # bare except
+            node.exc is None
+            # re-raised by name
+            or (isinstance(node.exc, ast.Name) and node.exc.id == self.except_name)
+            # new valid exception raised
+            or (
+                isinstance(node.exc, ast.Call)
+                and (
+                    (
+                        isinstance(node.exc.func, ast.Name)
+                        and node.exc.func.id == "BaseException"
+                    )
+                    or (
+                        isinstance(node.exc.func, ast.Attribute)
+                        and isinstance(node.exc.func.value, ast.Name)
+                        and node.exc.func.value.id == "trio"
+                        and node.exc.func.attr == "Cancelled"
+                    )
+                )
+            )
+        ):
+            # Error: something other than the exception was raised
+            self.problems.append(make_error(TRIO104, node.lineno, node.col_offset))
+        self.unraised = False
+        self.generic_visit(node)
+
+    def visit_Return(self, node: ast.Return):
+        if self.unraised:
+            # Error: must re-raise
+            self.problems.append(make_error(TRIO104, node.lineno, node.col_offset))
+        self.generic_visit(node)
+
+    def visit_Try(self, node: ast.Try):
+        if not self.unraised:
+            self.generic_visit(node)
+            return
+
+        # in theory it's okay if the try and all excepts re-raise,
+        # and there is a bare except
+        # but is a pain to parse and would require a special case for bare raises in
+        # nested excepts.
+        for n in (*node.body, *node.handlers, *node.orelse):
+            self.visit(n)
+            # re-set unraised to warn about returns in each block
+            self.unraised = True
+
+        # but it's fine if we raise in finally
+        for n in node.finalbody:
+            self.visit(n)
+
+    def visit_If(self, node: ast.If):
+        if not self.unraised:
+            self.generic_visit(node)
+            return
+
+        body_raised = False
+        for n in node.body:
+            self.visit(n)
+
+        # does body always raise correctly
+        body_raised = not self.unraised
+
+        self.unraised = True
+        for n in node.orelse:
+            self.visit(n)
+
+        # if body didn't raise, or it's unraised after else, set unraise
+        self.unraised = not body_raised or self.unraised
+
+    # disregard any raise's inside loops
+    def visit_For(self, node: ast.For):
+        outer_unraised = self.unraised
+        self.generic_visit(node)
+        self.unraised = outer_unraised
```

Comment:
Here (and in `visit_While`), we don't actually set `self.unraised`?

Reply:
```python
except trio.Cancelled:
  while [...]:
    if [...]:
      break
    raise
```
would set it and the `except` would think it's covered if it's not restored after the loop. Though I might have missed adding a test case for that

Reply:
Code comment to this effect please, to help me out next time!

Reply:
Yeah was a pretty terse comment - will try to comment more going forward!

Added a several comments to the other functions too ... catching a bug as I did :innocent: